src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs

```diff
@@ -263,15 +317,24 @@ protected override void VisitSwitchSection(BoundSwitchSection node, bool isLastS
             if (originalInputSlot <= 0)
             {
                 originalInputSlot = makeDagTempSlot(expressionType.ToTypeWithAnnotations(compilation), rootTemp);
-                initialState[originalInputSlot] = expressionType.State;
             }
+            Debug.Assert(originalInputSlot > 0);
+
+            // If the input of the switch (or is-pattern expression) is a tuple literal, we reuse the slots of
+            // those expressions (when possible), pretending that we are not copying them into a temporary ValueTuple instance
+            // to evaluate the patterns.  In this way we infer non-nullability of the original element's parts.
+            // We do not extend such courtesy to nested tuple literals.
+            var originalInputElementSlots = expression is BoundTupleExpression tuple
+                ? tuple.Arguments.SelectAsArray(a => MakeSlot(a))
```

Comment:
>SelectAsArray(a => MakeSlot(a)) [](start = 34, length = 31)

Consider passing in the walker: `SelectAsArray(static (a, w) => w.MakeSlot(a), this)`